mqtt/mqtt-broker/src/auth/authentication_edgehub.rs

```diff
@@ -0,0 +1,338 @@
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use serde_repr::Deserialize_repr;
+
+use crate::auth::{AuthId, Authenticator, Credentials};
+use crate::{AuthenticationError, Error};
+
+const API_VERSION: &str = "2020-04-20";
+
+pub struct EdgeHubAuthenticator {
+    url: String,
+}
+
+#[async_trait]
+impl Authenticator for EdgeHubAuthenticator {
+    type Error = Error;
+
+    async fn authenticate(
+        &self,
+        username: Option<String>,
+        credentials: Credentials,
+    ) -> Result<Option<AuthId>, Self::Error> {
+        self.authenticate_client((username, credentials).into())
+            .await
+            .map_err(Error::Authenticate)?
+            .into()
+    }
+}
+
+impl EdgeHubAuthenticator {
+    async fn authenticate_client(
+        &self,
+        request: EdgeHubAuthRequest,
+    ) -> Result<EdgeHubAuthResponse, AuthenticationError> {
+        let response = reqwest::Client::new()
+            .post(&self.url)
+            .json(&request)
+            .send()
+            .await
+            .map_err(|_| AuthenticationError::SendRequest)?;
+
+        if response.status() != reqwest::StatusCode::OK {
+            return Err(AuthenticationError::ProcessResponse);
+        }
+
+        let response = response
+            .json::<EdgeHubAuthResponse>()
+            .await
+            .map_err(|_| AuthenticationError::ProcessResponse)?;
+
+        if !response.version.eq(API_VERSION) {
```

Comment:
just curious why this check?

Reply:
I added version to the record structure, and the server also checks it (and sends it back). The check is for avoid and update at server side but not following at client side.